python/modules/IcePy/Slice.cpp

```diff
@@ -72,153 +65,76 @@ IcePy_loadSlice(PyObject* /*self*/, PyObject* args)
     {
-        vector<string> optargs = opts.argVec("U");
-        for (const auto& arg : optargs)
-        {
-            cppArgs.push_back("-U" + arg);
-        }
+        cppArgs.push_back("-U" + arg);
     }
-    if (opts.isSet("I"))
+
+    for (const auto& path : opts.argVec("I"))
     {
-        includePaths = opts.argVec("I");
-        for (const auto& path : includePaths)
-        {
-            cppArgs.push_back("-I" + path);
-        }
+        cppArgs.push_back("-I" + path);
     }
+
     debug = opts.isSet("d") || opts.isSet("debug");
-    all = opts.isSet("all");
 
-    for (const auto& file : files)
+    vector<PythonCodeFragment> fragments;
+    try
     {
-        Slice::PreprocessorPtr icecpp = Slice::Preprocessor::create("icecpp", file, cppArgs);
-        FILE* cppHandle = icecpp->preprocess(true, "-D__SLICE2PY__");
-
-        if (!cppHandle)
-        {
-            PyErr_Format(PyExc_RuntimeError, "Slice preprocessing failed for `%s'", cmd);
-            return nullptr;
-        }
-
-        UnitPtr u = Slice::Unit::createUnit("python", Slice::Python::pyLinkFormatter, all);
-        int parseStatus = u->parse(file, cppHandle, debug);
-
-        if (!icecpp->close() || parseStatus == EXIT_FAILURE)
-        {
-            PyErr_Format(PyExc_RuntimeError, "Slice parsing failed for `%s'", cmd);
-            u->destroy();
-            return nullptr;
-        }
-
-        //
-        // Generate the Python code into a string stream.
-        //
-        ostringstream codeStream;
-        IceInternal::Output out(codeStream);
-        out.setUseTab(false);
+        fragments = Python::dynamicCompile(files, cppArgs, debug);
+    }
+    catch (const std::exception& ex)
+    {
+        PyErr_Format(PyExc_RuntimeError, "error during Slice compilation: %s", ex.what());
+        return nullptr;
+    }
 
-        // TODO rework loadSlice
-        // generate(u, all, includePaths, out);
-        if (u->getStatus() == EXIT_FAILURE)
+    for (const auto& fragment : fragments)
+    {
+        PyObject* moduleRef = PyImport_AddModule(fragment.moduleName.c_str());
+        if (!moduleRef)
         {
-            PyErr_Format(PyExc_RuntimeError, "Slice validation failed for `%s'", cmd);
-            u->destroy();
+            PyErr_Format(PyExc_RuntimeError, "failed to add module `%s'", fragment.moduleName.c_str());
             return nullptr;
         }
-        u->destroy();
-
-        string code = codeStream.str();
-
-        //
-        // We need to invoke Ice.updateModules() so that all of the types we've just generated
-        // are made "public".
-        //
-        code += "\nIce.updateModules()\n";
 
-        PyObjectHandle src{
-            Py_CompileString(const_cast<char*>(code.c_str()), const_cast<char*>(file.c_str()), Py_file_input)};
-        if (!src.get())
+        // The file name is set to the Slice file the code was generated from, or to the package index file name for
+        // package index fragments.
```

Comment:
The package index is created from multiple Slice files, that is why we don't use a Slice file as the Py_file_input.